### PR TITLE
add benchmark for knn

### PR DESCRIPTION
**Pull request by leewyang**
Based on [original code](https://github.com/lijinf2/spark-rapids-ml/commit/e02892a0d4facfae6b02686e48b6e9f7741fb5cd) from @lijinf2.

Note: `CumlContext.get_ifname_from_ip()` wasn't working on my single-node dev env (and inside the build docker image).  I added a try block around it to just leave `UCXPY_IFNAME` unset on any errors (per suggestion from @eordentlich), which seems to work fine.

### CODE REVIEW

**Comment by leewyang** [comment]
build

**Comment by eordentlich** [comment]
build

File: benchmark/benchmark/bench_nearest_neighbors.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    def run_once(
+        self,
+        spark: SparkSession,
+        train_df: DataFrame,
+        features_col: Union[str, List[str]],
+        transform_df: Optional[DataFrame],
+        label_name: Optional[str],
+    ) -> Dict[str, Any]:
+        """
+        This function evaluates the runtimes of SparkCuml NearestNeighbors and Spark LSH, but
```
Comment (lijinf2) [inline benchmark/benchmark/bench_nearest_neighbors.py:49]: I forgot to replace "SparkCuml" with "SparkRapidsML".

Comment (leewyang) [reply benchmark/benchmark/bench_nearest_neighbors.py:49]: Missed that too... will update.

File: benchmark/benchmark/bench_nearest_neighbors.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        func_start_time = time.time()
+
+        first_col = train_df.dtypes[0][0]
+        first_col_type = train_df.dtypes[0][1]
+        is_array_col = True if "array" in first_col_type else False
+        is_vector_col = True if "vector" in first_col_type else False
+        is_single_col = is_array_col or is_vector_col
+        if not is_single_col:
+            input_cols = [c for c in train_df.schema.names]
+        output_col = "cluster_idx"
```
Comment (lijinf2) [inline benchmark/benchmark/bench_nearest_neighbors.py:67]: This variable can be deleted.

File: benchmark/benchmark/bench_nearest_neighbors.py (hunk 1 above)
Comment (leewyang) [reply benchmark/benchmark/bench_nearest_neighbors.py]: done.

**Comment by lijinf2** [comment]
build

**Comment by lijinf2** [comment]
build

**Comment by lijinf2** [comment]
LGTM, meanwhile it is unclear how to set bucketLegnth and numHashTables properly for CPU LSH. 
Created a JIR ticket to track this issue: https://jirasw.nvidia.com/browse/DML-468.

File: src/spark_rapids_ml/common/cuml_context.py [hunk 3]
```diff
@@ -81,9 +81,12 @@ def __init__(
             # the code occasionally run fail without setting the variables
             # TODO: will have to figure how to make this more flexible to take advantage of higher speed interconnects on multi-gpu nodes
             my_ip = self._ips[self._rank]
-            my_ifname = CumlContext.get_ifname_from_ip(my_ip)
             os.environ["UCX_TLS"] = "tcp,cuda_copy,cuda_ipc"
-            os.environ["UCXPY_IFNAME"] = my_ifname
+            try:
+                my_ifname = CumlContext.get_ifname_from_ip(my_ip)
```
[the remark this answers is not in the document]
Comment (leewyang) [reply src/spark_rapids_ml/common/cuml_context.py]: It doesn't print any errors.

File: benchmark/benchmark/bench_nearest_neighbors.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                "cpu fit time", lambda: cpu_estimator.fit(vector_df)
+            )
+
+            def cpu_transform(
+                model: BucketedRandomProjectionLSHModel, df: DataFrame, n_neighbors: int
+            ) -> None:
+                queries = df.collect()
+                for row in queries:
+                    query = row[first_col]
+                    knn_df = model.approxNearestNeighbors(
```
[the remark this answers is not in the document]
Comment (leewyang) [reply benchmark/benchmark/bench_nearest_neighbors.py:154]: Not sure, but I'd suggest deferring this to future work.

**Comment by leewyang** [comment]
build

File: benchmark/benchmark/bench_nearest_neighbors.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lijinf2) [reply benchmark/benchmark/bench_nearest_neighbors.py:154]: Added this comment to the ticket https://jirasw.nvidia.com/browse/DML-468 for reference.

**Comment by lijinf2** [comment]
build

**Review by lijinf2 — APPROVED** [review]
Looks nice. Thank you!

**Pull request merged**